design/service-apis-implementation.md

```diff
@@ -0,0 +1,229 @@
+If this field is empty, then no GatewayClasses match, there is no active GatewayClass, and no Gateways will be checked.
+
+Note that Contour's status update machinery should ensure that only status updates that are changes are actually applied.
+
+A DAG rebuild should only be kicked off if the GatewayClass value in the KubernetesCache changed.
+
+### Gateway
+
+When ingesting Gateways, Contour will import all Gateways into the KubernetesCache.
+
+#### Validity
+
+Before processing Gateways, Contour will run a quick check across all gateways to look for conflicting or out-of-scope Listeners.
+
+Contour will use the `GatewayClassDetails` field in the Kubernetes cache to check each object's `spec.gatewayClassName` field against the GatewayClass name.
+If they do not match, the Gateway will be skipped from processing.
+A debug-level log line should be output here to indicate the skip.
+
+Listener **conflict** is defined as two listeners sharing 
+Listeners that cannot be merged *within a Gateway* are invalid and will be rejected, and their status updated accordingly.
+
+When Listeners cannot be merged *across Gateways* the **oldest** Gateway's Listeners (by creation timestamp) will be accepted, and other Listeners will be rejected.
+An accepted Listener will also have a condition placd on it to indicate the Gateway that contains Listeners in conflict with it.
+
+This behavior is similar to the HTTPProxy processor's `validHTTPProxies()` method.
+
+#### Listener merging rules
+
+TODO: The basic principle is to allow for people to specify lots of names across a set of Gateways, or within one Gateway.
+
+Listeners that match on ProtocolType and PortNumber can be merged, using Hostname as a discriminator.
+
+TODO: There are some edge cases to check into here because TLS config can be specified both here and in the Route.
+Also need to check for if you can mix-and-match between Routes and here, using here as a fallback?
+
+### HTTPRoute
+
+### TLSRoute
+
+
+
+### Other concerns
+
+TODO: What do we do about updating status information? Probably just rely on whatever provisions Envoy to put the reachability info in the Gateway Status.
+
+
+For each object, I need to specify:
+How the watch will work
+Where the integration will be plumbed
+How do we handle the inter-related dynamic configuration?
+
+A detailed design describing how the changes to the product should be made.
+
+The names of types, fields, interfaces, and methods should be agreed on here, not debated in code review.
+The same applies to changes in CRDs, YAML examples, and so on.
+
+Ideally the changes should be made in sequence so that the work required to implement this design can be done incrementally, possibly in parallel.
+
+## Alternatives Considered
+TODO: This will need further explanation of what the solution would look like if we didn't make one GatewayClass == One Contour.
```

Comment:
The contour-operator [design doc](https://github.com/projectcontour/contour-operator/pull/171/files) more closely aligns with this alternative approach. An instance of Contour (i.e. Contour Deployment, Envoy DaemonSet, etc.) equals one Gateway. GatewayClass is simply a resource that points to an implementation-specific config object, i.e. `ContourConfig`. Contour will only be concerned with xRoute resources and using `xroute.spec.gateways[]` to know if it's responsible for handling the route.

Reply:
I think we cleared this up in a maintainer's meeting, now that the Contour installation (= Contour deployment and Envoy DaemonSet and associated bits) is 1-to-1 with a named Gateway.